refurbished/feedback.py

```diff
@@ -14,27 +17,55 @@ def __init__(self, format):
     def echo(self, text, nl=True, err=False):
         click.echo(text, nl=nl, err=err)
 
-    def result(self, result):
+    def result(self, result) -> None:
         if self.format == "text":
             click.echo(
                 result.str(),
-                # delefate newlines to the result class
+                # delegate newlines to the result class
                 nl=False,
             )
         elif self.format == "json":
+            # assumption: entries are all pydantic dataclasses
             click.echo(
                 json.dumps(result.data(), indent=2, default=pydantic_encoder),
                 # delegate newline to json.dumps
                 nl=False,
             )
         elif self.format == "ndjson":
-            for product in result.data():
+            # assumption: entries are all pydantic dataclasses
+            for entry in result.data():
                 click.echo(
-                    json.dumps(product, default=pydantic_encoder),
+                    json.dumps(entry, default=pydantic_encoder),
                     # The newline is required by the format to separate the
                     # JSON objects
                     nl=True,
                 )
+        elif self.format == "csv":
+            entries = result.data()
+
+            # we need at least one entry to get the fieldnames from
+            # the pydantic dataclass for the csv header
+            if len(entries) == 0:
+                return
+
+            # assumption: entries are all pydantic dataclasses
+            if not is_dataclass(entries[0]):
+                raise TypeError(
+                    "CSV format requires a list of dataclasses, got "
+                    f"{type(entries[0])}"
+                )
+
+            fieldnames = (
+                entries[0].__pydantic_model__.schema()["properties"].keys()
```

Comment:
Move this into a ProductsResult helper instead (or even upper in the stack) so you don't leak the implementation outside?